pkg/controller/machineset/machineset_controller_test.go

```diff
@@ -0,0 +1,275 @@
+var machinesetKey2 = types.NamespacedName{Name: "machineset2", Namespace: "default"}
+
+const timeout = time.Second * 10
+
+func TestScale(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	rawProviderSpec, err := json.Marshal(&bmv1alpha1.BareMetalMachineProviderSpec{
+		HostSelector: bmv1alpha1.HostSelector{
+			MatchLabels: map[string]string{"size": "large"},
+		},
+	})
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	instance := &machinev1beta1.MachineSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "machineset1",
+			Namespace:   "default",
+			Annotations: map[string]string{AutoScaleAnnotation: "yesplease"},
+		},
+		Spec: machinev1beta1.MachineSetSpec{
+			Template: machinev1beta1.MachineTemplateSpec{
+				Spec: machinev1beta1.MachineSpec{
+					ProviderSpec: machinev1beta1.ProviderSpec{
+						Value: &runtime.RawExtension{Raw: rawProviderSpec},
+					},
+				},
+			},
+			Selector: metav1.LabelSelector{
+				MatchLabels: map[string]string{"machine.openshift.io/cluster-api-machineset": "cluster0-worker"},
+			},
+		},
+	}
+	// machine1 has a different label than the MachineSet's Selector, so its
+	// consumed host should not be counted as part of that MachineSet's
+	// potential hosts.
+	machine1 := machinev1beta1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "machine1",
+			Namespace: "default",
+			Labels:    map[string]string{"machine.openshift.io/cluster-api-machineset": "cluster0-storage"},
+		},
+	}
+	// machine2 has a label that matches the MachineSet's Selector, so its
+	// consumed host should be counted as part of that MachineSet's potential
+	// hosts even if that BareMetalHost does not otherwise match.
+	machine2 := machinev1beta1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "machine2",
+			Namespace: "default",
+			Labels:    map[string]string{"machine.openshift.io/cluster-api-machineset": "cluster0-worker"},
+		},
+	}
+	host1 := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "host1",
+			Namespace: "default",
+			Labels:    map[string]string{"size": "large"},
+		},
+	}
+	// This host has a different label, but its consuming Machine is
+	// part of the MachineSet, so it should be counted.
+	host2 := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "host2",
+			Namespace: "default",
+			Labels:    map[string]string{"size": "small"},
+		},
+		Spec: bmh.BareMetalHostSpec{
+			ConsumerRef: &v1.ObjectReference{
+				Kind:       "Machine",
+				APIVersion: machinev1beta1.SchemeGroupVersion.String(),
+				Name:       "machine2",
+				Namespace:  "default",
+			},
+		},
+	}
+	// This host has a different label, so it should not match
+	host3 := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "host3",
+			Namespace: "default",
+			Labels:    map[string]string{"size": "extramedium"},
+		},
+	}
+	// This host is consumed by a Machine in a different MachineSet, so it
+	// should not be counted
```

Comment:
Since the reconcile loop tests for it, it would be good to add a host that's consumed by something other than a Machine.